Bailing, prefix arguments, and command cycling

tad-ashlock opened the pull request:
Everyone I know who uses evil-god-state has had problems with getting stuck in it. Until we found the 'bail' code, we were ready to give up on evil-god-state. For that reason, I think it should be part of the base package.

Evil-god-state was aborting whenever I hit 'u' (for Ctrl-u).  This made prefix arguments impossible.  The trick is to detect when prefix arguments are being added and stay in evil-god-state.

With the command cycling change, Org Mode is easier to use with evil-god-state.  I leave <TAB> (Ctrl-i) for jumping forward, then use evil-god-state to run org-cycle with "SPC i".  But this only works if evil-god-state doesn't change the value of the 'last-command' variable.

tad-ashlock committed:
Configure <ESC> to always abort evil-god-state

This change posted by valurs at
https://gist.github.com/gridaphobe/9765143#comment-1222831.
It is possible to get stuck in evil-god-state.  This change makes
getting out of it easy.

tad-ashlock committed:
Add support for prefix command arguments

tad-ashlock committed:
Add support for command cycling

Some commands change their behavior based on the previous command
executed.  (For example, Org Mode's org-cycle command.)  This
change hides the Evil-God commands so that command cycling works.

gridaphobe commented on evil-god-state.el:
```diff
@@ -98,5 +113,14 @@
+  (evil-stop-execute-in-god-state)
+  (evil-god-stop-hook)
+  (evil-normal-state))
+(evil-define-key 'god global-map [escape] 'evil-god-state-bail)
```
I'd prefer to not define any keybindings in this file, letting users make that decision for themselves. Could you move the keybinding up to the commentary section where I have the other recommendations, and perhaps add it to the README as well?

gridaphobe commented:
Thanks for the PR! Haven't tested it yet but everything looks sensible to me.

I'm especially grateful for the bailing code, I got stuck in god-state a lot when I first threw this together too.. But by the time I posted the code it had stopped happening altogether, so I thought it wasn't an issue any more. Regardless, it's good to have a failsafe :)

tad-ashlock committed:
Make the `evil-god-state-bail' keybinding a documented suggestion

Remove the `evil-god-state-bail' keybinding from the code and put it
into the commentary section with the other recommendations, and also
put it in the README file.

tad-ashlock commented:
Thanks for the Emacs package!  It's one of those that I'm using every day
now.

I've added another commit to move the keybinding out of the code.  That is
a good call, by the way.  This is the first time I've ever used GitHub and
did a pull request.  So forgive me if I screw things up.  Also, pointers
are very welcome.

Thanks!
--Tad

On Sat, Aug 23, 2014 at 10:36 AM, Eric Seidel notifications@github.com
wrote:

> Thanks for the PR! Haven't tested it yet but everything looks sensible to
> me.
> 
> I'm especially grateful for the bailing code, I got stuck in god-state a
> lot when I first threw this together too.. But by the time I posted the
> code it had stopped happening altogether, so I thought it wasn't an issue
> any more. Regardless, it's good to have a failsafe :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/gridaphobe/evil-god-state/pull/1#issuecomment-53158023
> .

gridaphobe commented:
Ok, everything seems to work great, thanks again!
